### ISSUE TITLE
`DYN`: don't issue a hash request

### ISSUE THREAD

**Issue by plafer** [0xPolygonMiden/miden-vm#1523]
It seems like `DYN` could not issue any hash request, and also not have an associated `END`, since we're only hashing dummy values? So `DYN` would still populate the block hash table, but have no other effect on the trace.

_Originally posted by @plafer in https://github.com/0xPolygonMiden/miden-vm/pull/1516#discussion_r1781552482_

**Comment by bobbinth** [comment]
To clarify, is the issue here that we could optimize things a bit more by not filling the hasher chiplet with rows to compute hash of of zeros?

**Comment by plafer** [comment]
Exactly. We will basically save 8 hasher chiplet rows per `DYN` (and 1 main trace row from not having a corresponding `END`)

**Comment by plafer** [comment]
Also, I believe `DYN` could be made an operation that is included in basic blocks then? 

In #1091, we need the decoder helper registers to save the word loaded from memory (for the chiplets bus and block hash table). However, technically, since `DYN` is currently a control block, we [shouldn't be using its helper registers](https://github.com/0xPolygonMiden/miden-vm/blob/f92a4633b07c58ed70236710259d243bca1918ff/processor/src/decoder/mod.rs#L658).

**Comment by bobbinth** [comment]
> Exactly. We will basically save 8 hasher chiplet rows per `DYN` (and 1 main trace row from not having a corresponding `END`)

Makes sense. One other thing we can try to do in the future is to use multiplicities somehow in the hasher chiplet so that we would not need to have a trace for computing the same hash more than once.

> I believe `DYN` could be made an operation that is included in basic blocks then?

This would be tricky as basic blocks don't have children and `DYN` could invoke an arbitrary block (e..g, a `JOIN` block). So, we'd need to end the current basic block, then execute the dynamic block, and then execute the rest basic block - but this would be basically the same thing as having `DYN` be its own node (i.e., not in basic block).

**Comment by bobbinth** [comment]
@plafer - is this still open?

**Comment by plafer** [comment]
Yes this was never addressed

**Comment by plafer** [comment]
The proposed solution ends up creating a lot more problems than I originally intended. The main issue with not issuing a hash request is that the chiplets bus currently always issues a hash response message on `END`. Therefore we must now detect that the `END` is associated with a `DYN` with a flag, and *not* issue a hash request when that flag is set. The main problem is that we're out of "flag space" on `END`, currently `is_loop_body`, `is_loop`, `is_call` and `is_syscall`. 

There's also the issue that since we no longer issue a hash request, we no longer get a fresh `addr` for the `DYN` block, i.e. the `DYN` block has the same address as the parent (which must also be handled as a special case when restoring the `addr` on `END` - and I believe 1 or 2 constraints to make sure the `addr` doesn't change on `DYN`). I couldn't come up with specific attacks that exploit this directly, but regardless it adds complexity to remember that `DYN` nodes don't get a new address, making our system more bug-prone IMO.

TLDR, I think the complexity cost of implementing that change isn't worth the few hasher chiplet rows we would save by saving 1 hash request per DYN node (which there aren't typically that many), and we should probably close as "won't do".

**Comment by bobbinth** [comment]
Makes sense. Agreed that the extra complexity is probably not worth it, and AFAICT, this is purely an optimization, right? (i.e., there are not soundness issues with the current approach).

I'm fine with closing this as "won't do" - but would be good to get @Al-Kindi-0 thoughts too.

**Comment by Al-Kindi-0** [comment]
I think that with the recent hasher chiplet refactoring the overhead is really small. More precisely: the 16-row permutation cycle is paid only once per distinct `DYN` domain (once for `dynexec` and once for `dyncall`), regardless of how many `DYN` nodes the program executes. The only cost that scales linearly with the number of `DYN` nodes is the two rows in the hasher controller. In my opinion this should be acceptable when we take all factors into consideration, so I'm in favor of closing as "won't do".
